<a href="https://colab.research.google.com/github/Nikolai-N484/Data201_NikolaiN/blob/main/Project3/Nikolai_Navarro_Project3_Week14.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Data 201 Final Project**
## **Nikolai Navarro**
## Spring 2026
---

## **Introduction**
Write a short introduction describing:
- Your dataset
- Your prediction goal
- Why the topic is interesting or important
- Possible limitations or bias in the dataset

Include a citation for your dataset source.

I will be reusing the same data set I used for project 1. This dataset was collected by Montgomery County via its Automated Crash Reporting System which includes reports from Maryland state police such as Montgomery County Police, Gaithersburg Police, Rockville Police, or the Maryland-National Capital Park Police. This dataset focuses on traffic collisions involving non-motorists which mean pedestrians and cyclists.

The data provided is updated weekly and may include missing information due to mechanical or human error. Some information may also be missing because of ongoing investigations.

I want to predict the crash type/severity of the crash using variables from the data set such as the lighting the weather and the road conditions. I chose this topic because it has data about crashes that occur in our county and being able to see what factors play a large role in determine the types of crashes can help us to understand what we can try to do to prevent these crashes from happening again in the future.

**Citation:**\
“Crash Reporting - Non-Motorists Data: Open Data Portal.” Crash Reporting - Non-Motorists Data | Open Data Portal, Montgomery County, 27 Mar. 2026, data.montgomerycountymd.gov/Public-Safety/Crash-Reporting-Non-Motorists-Data/n7fk-dce5/about_data.

---

## **Dataset and Preparation**

Include:
- Dataset size
- Explanation of variables
- Missing value handling
- Encoding categorical variables if needed
- Train/test split

Include at least one table describing variables.

In [53]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt

df = pd.read_csv("https://raw.githubusercontent.com/Nikolai-N484/Data201_NikolaiN/refs/heads/main/Crash_Reporting_-_Non-Motorists_Data_20260329.csv")

df.head()

,Report Number,Local Case Number,Agency Name,ACRS Report Type,Crash Date/Time,Route Type,Road Name,Cross-Street Name,Off-Road Description,Municipality,...,Pedestrian Type,Pedestrian Movement,Pedestrian Actions,Pedestrian Location,At Fault,Injury Severity,Safety Equipment,Latitude,Longitude,Location
0,MCP3280007Y,260012513,MONTGOMERY,Injury Crash,03/22/2026 07:30:00 PM,County Route,STEWART LA,OLD COLUMBIA PIKE OLD COLUMBIA PIKE (SB/L),NaN,NaN,...,Pedestrian,Cross/Enter at Intersection,None (No Improper Action),Intersection Marked Crosswalk,No,Possible Injury,NaN,39.046210,-76.986543,"(39.04620981, -76.98654308)"
1,MCP3126007Y,260012218,MONTGOMERY,Injury Crash,03/20/2026 08:01:00 PM,NaN,NaN,NaN,Parking Lot Way MILESTONE SHOPPING CENT...,NaN,...,Pedestrian,Cross/Enter at Intersection,None (No Improper Action),Intersection Marked Crosswalk,No,Suspected Minor Injury,NaN,39.199092,-77.247840,"(39.1990915, -77.24783967)"
2,MCP3126007Y,260012218,MONTGOMERY,Injury Crash,03/20/2026 08:01:00 PM,NaN,NaN,NaN,Parking Lot Way MILESTONE SHOPPING CENT...,NaN,...,Pedestrian,Cross/Enter at Intersection,None (No Improper Action),Intersection Marked Crosswalk,No,No Apparent Injury,NaN,39.199092,-77.247840,"(39.1990915, -77.24783967)"
3,MCP3077002M,260012176,MONTGOMERY,Injury Crash,03/20/2026 03:53:00 PM,County Route,FATHER HURLEY BLVD,NaN,NaN,NaN,...,Pedestrian,Cross/Enter at Intersection,None (No Improper Action),Intersection Marked Crosswalk,No,Possible Injury,NaN,39.191983,-77.272170,"(39.19198337, -77.27216959)"
4,MCP30790088,260011936,MONTGOMERY,Injury Crash,03/19/2026 09:19:00 AM,NaN,NaN,NaN,Parking Aisle PARKING LOT IN FRONT OF L...,NaN,...,Pedestrian,Other,None (No Improper Action),Other,No,Possible Injury,NaN,39.179210,-77.260974,"(39.17920958, -77.26097442)"


In [54]:
shape = df.shape
print(shape)

(7147, 29)


In [55]:
df["Crash_Month"] = pd.to_datetime(df["Crash Date/Time"]).dt.month
df["Crash_Day"] = pd.to_datetime(df["Crash Date/Time"]).dt.day
df["Crash_Year"] = pd.to_datetime(df["Crash Date/Time"]).dt.year
df["Crash_Time"] = pd.to_datetime(df["Crash Date/Time"]).dt.time

df.head()

/tmp/ipykernel_2233/2059507749.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df["Crash_Month"] = pd.to_datetime(df["Crash Date/Time"]).dt.month
/tmp/ipykernel_2233/2059507749.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df["Crash_Day"] = pd.to_datetime(df["Crash Date/Time"]).dt.day
/tmp/ipykernel_2233/2059507749.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df["Crash_Year"] = pd.to_datetime(df["Crash Date/Time"]).dt.year
/tmp/ipykernel_2233/2059507749.py:4: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `

,Report Number,Local Case Number,Agency Name,ACRS Report Type,Crash Date/Time,Route Type,Road Name,Cross-Street Name,Off-Road Description,Municipality,...,At Fault,Injury Severity,Safety Equipment,Latitude,Longitude,Location,Crash_Month,Crash_Day,Crash_Year,Crash_Time
0,MCP3280007Y,260012513,MONTGOMERY,Injury Crash,03/22/2026 07:30:00 PM,County Route,STEWART LA,OLD COLUMBIA PIKE OLD COLUMBIA PIKE (SB/L),NaN,NaN,...,No,Possible Injury,NaN,39.046210,-76.986543,"(39.04620981, -76.98654308)",3,22,2026,19:30:00
1,MCP3126007Y,260012218,MONTGOMERY,Injury Crash,03/20/2026 08:01:00 PM,NaN,NaN,NaN,Parking Lot Way MILESTONE SHOPPING CENT...,NaN,...,No,Suspected Minor Injury,NaN,39.199092,-77.247840,"(39.1990915, -77.24783967)",3,20,2026,20:01:00
2,MCP3126007Y,260012218,MONTGOMERY,Injury Crash,03/20/2026 08:01:00 PM,NaN,NaN,NaN,Parking Lot Way MILESTONE SHOPPING CENT...,NaN,...,No,No Apparent Injury,NaN,39.199092,-77.247840,"(39.1990915, -77.24783967)",3,20,2026,20:01:00
3,MCP3077002M,260012176,MONTGOMERY,Injury Crash,03/20/2026 03:53:00 PM,County Route,FATHER HURLEY BLVD,NaN,NaN,NaN,...,No,Possible Injury,NaN,39.191983,-77.272170,"(39.19198337, -77.27216959)",3,20,2026,15:53:00
4,MCP30790088,260011936,MONTGOMERY,Injury Crash,03/19/2026 09:19:00 AM,NaN,NaN,NaN,Parking Aisle PARKING LOT IN FRONT OF L...,NaN,...,No,Possible Injury,NaN,39.179210,-77.260974,"(39.17920958, -77.26097442)",3,19,2026,09:19:00


In [59]:
df = df[["ACRS Report Type", "Collision Type", "Weather", "Surface Condition", "Light", "Traffic Control", "Injury Severity", "Crash_Month", "Crash_Time"]]
df

,ACRS Report Type,Collision Type,Weather,Surface Condition,Light,Traffic Control,Injury Severity,Crash_Month,Crash_Time
0,Injury Crash,Single Vehicle,Clear,Dry,Daylight,Stop Sign,Possible Injury,3,19:30:00
3,Injury Crash,Single Vehicle,Clear,Dry,Daylight,Lane Use Control Signal,Possible Injury,3,15:53:00
6,Injury Crash,Single Vehicle,Clear,Dry,Dusk,Pedestrian Crossing Sign,Suspected Minor Injury,3,18:30:00
7,Injury Crash,Single Vehicle,Clear,Dry,Daylight,Pedestrian Crossing Sign,Suspected Minor Injury,3,10:35:00
8,Injury Crash,Single Vehicle,Clear,Dry,Daylight,Traffic Control Signal,Suspected Minor Injury,3,15:03:00
...,...,...,...,...,...,...,...,...,...
7137,Injury Crash,OTHER,CLEAR,DRY,DAYLIGHT,TRAFFIC SIGNAL,SUSPECTED MINOR INJURY,1,14:56:00
7139,Injury Crash,SINGLE VEHICLE,SNOW,SLUSH,DARK LIGHTS ON,NO CONTROLS,POSSIBLE INJURY,1,18:35:00
7140,Injury Crash,SINGLE VEHICLE,SNOW,ICE,DAYLIGHT,STOP SIGN,SUSPECTED SERIOUS INJURY,1,08:04:00
7141,Fatal Crash,OTHER,CLEAR,DRY,DARK LIGHTS ON,TRAFFIC SIGNAL,FATAL INJURY,1,17:55:00


In [57]:
na = df.isna().sum()
print(na)

ACRS Report Type        0
Collision Type        178
Weather               458
Surface Condition    1691
Light                  46
Traffic Control      1097
Injury Severity         0
Crash_Month             0
Crash_Time              0
dtype: int64


In [58]:
df = df.dropna()
shape2 = df.shape
print(shape2)

na2 = df.isna().sum()
print(na2)

(4686, 9)
ACRS Report Type     0
Collision Type       0
Weather              0
Surface Condition    0
Light                0
Traffic Control      0
Injury Severity      0
Crash_Month          0
Crash_Time           0
dtype: int64


After cleaning the data set to remove any NA's, selecting certain variables, creating new variables to account for the time and month the data set ended up with a total of 4686 rows and 9 columns.

---
## **Exploratory Data Analysis**

Create:
- **2 meaningful visualizations**
- Short interpretations

Discuss:
- Interesting patterns
- Trends
- Possible relationships

Keep this section concise.

---
## **Decision Tree Analysis (Core Section)**

This is the most important section of the project.

You must:
- Train a Decision Tree Regressor
- Try at least **two different tree depths**
- Briefly discuss possible overfitting
- Include a visualization or diagram of your tree

Also explain:
- The root node
- One important split
- One example prediction path

---
## **Ensemble Models**

Train:
- Random Forest
- Gradient Boosting

Compare them to your single decision tree.

Discuss:
- Did performance improve?
- Which model generalized better?
- Which model was easier to explain?

You are NOT expected to perform advanced hyperparameter tuning.

---
## **Model Evaluation**

You must include:

| Metric | Meaning |
|---|---|
| MSE | Lower is better |
| \(R^2\) | Higher is better |

Include:
- A comparison table
- Actual vs Predicted visualization
- Short interpretation of results

---
## **Feature Importance**

For at least one ensemble model:
- Display feature importance
- Identify the top 3 most important features

Important:
- Feature importance does NOT prove causation

---
## **Final Reflection: The Model Battle**

Answer the following questions:

1. Which model performed best?
2. Which model was easiest to explain?
3. Which model would you choose in a real-world setting and why?